packages/gallery/src/components/gallery/proGallery/slideshowView.js

```diff
@@ -337,35 +341,14 @@ class SlideshowView extends GalleryComponent {
 
     this.isAutoScrolling = true;
 
-    if (isAutoTrigger) {
-      // ---- Called by the Auto Slideshow ---- //
-      if (this.isLastItem()) {
-        // maybe this should be isLastItemFullyVisible now that we have both. product- do we allow autoSlideshow in other layouts ( those that could have more than one item displayed in the galleryWidth)
-        currentGroup = 0;
-        scrollDuration = 0;
-      }
-    } else {
-      // ---- Called by the user (arrows, keys etc.) ---- //
-      // this.startAutoSlideshowIfNeeded(this.props.styleParams);
-      const scrollingPastLastItem =
-        (direction >= 1 && this.isLastItem()) ||
-        (direction <= -1 && this.isFirstItem());
-      if (scrollingPastLastItem) {
-        this.isSliding = false;
-        return;
-      }
-    }
     // ---- navigate ---- //
     try {
-      const isScrollingPastEdge =
-        !isAutoTrigger &&
-        ((direction >= 1 && this.isLastItemFullyVisible()) ||
-          (direction <= -1 && this.isFirstItemFullyVisible()));
       const scrollMarginCorrection = this.getStyles().margin || 0;
       const _scrollDuration =
         scrollDuration || this.props.styleParams.scrollDuration || 400;
+      const shouldAllowScroll = !this.shouldNotAllowScroll({ scrollingUpTheGallery });
 
-      !isScrollingPastEdge &&
+      shouldAllowScroll &&
```

Comment:
The entire code in lines 278-329 (**lets call it code A**) is duplicated between the lines and 345-379 (**lets call it code B**). 

Here is the list of  **differences** between these two chunks of code is (**everything else is duplicated**)

1. Code A calls `scrollToItem`. Code B calls `scrollToGroup`. (Note that the arguments passed there are **also duplicated**)
2. There's the `this.props.styleParams.groupSize === 1` handling only in code A
3. The `activeIndex` has different values between code A and code B
4. There's an `ignoreScrollPosition` handling in code A
5. The `console.error` has different text in the `catch` block

Reply:
Hey Yakov & Nir,
This refactor was about the block of next and the block of the scroll itself in two functions, this is why Yakov did not touch other possible refactors.

Yakov - let's finish this up with Nirs list to leave a really clean code.